WMF Framework/EventLoggingService.swift

```diff
@@ -286,18 +127,39 @@ public class EventLoggingService : NSObject, URLSessionDelegate {
             DDLogDebug("EventLoggingService: \(record.objectID) recorded!")
             
             self.save()
-
-            if self.shouldSendImmediately {
-                self.tryPostEvents()
-            }
         }
     }
     
     @objc
-    private func tryPostEvents() {
+    private func tryPostEvents(_ completion: (() -> Void)? = nil) {
         let operation = AsyncBlockOperation { (operation) in
             let moc = self.managedObjectContext
             moc.perform {
+                let pruneFetch = NSFetchRequest<NSFetchRequestResult>(entityName: "WMFEventRecord")
+                pruneFetch.returnsObjectsAsFaults = false
+                
+                let pruneDate = Date().addingTimeInterval(-(self.pruningAge)) as NSDate
+                pruneFetch.predicate = NSPredicate(format: "(recorded < %@) OR (posted != nil) OR (failed == TRUE)", pruneDate)
+                let delete = NSBatchDeleteRequest(fetchRequest: pruneFetch)
```

Comment:
What's the reason for choosing `NSBatchDeleteRequest` over `wmf_batchProcessObjects` (as used in `ReadingListsSyncOperation` here https://github.com/wikimedia/wikipedia-ios/blob/b726e6603fe274d8748206b2d82727be54f5e069/WMF%20Framework/ReadingListsSyncOperation.swift#L255)?

Reply:
`NSBatchDeleteRequest` is more efficient as it deletes the items without populating them into memory. However, if I'm remembering correctly, it doesn't process object relationship changes or send object change notifications (if you have the deleted objects in memory already, they're invalid). In this case, since we're only ever fetching the events and not relying on change notifications, it's fine. For the reading lists, we want to make sure the related objects are updated and that we get change notifications. For what it's worth, this change is just moving the event prune code that was in existence, not a new implementation.

Reply:
Makes sense, thanks 👍